File: DzBase/permission/src/main/java/com/daipi/permission/aspectversion/MainActivity.java
package com.daipi.permission.aspectversion;

import android.Manifest;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.daipi.permission.R;
import com.daipi.permission.aspectversion.annotation.Permission;
import com.daipi.permission.aspectversion.annotation.PermissionCancel;
import com.daipi.permission.aspectversion.annotation.PermissionDenied;

/**
 * TODO 用户的角度上  use 我们写的权限申请库
 *
 * 你是架构师设计，你的组员用户，只需要面对三个注解就行了
 */
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // 不用反射，也不用 APT

        /*Xutils.xxx(this);  反射
        Dagger2.xxxx(this);  APT
        ARouter.xxx(this);  APT
        */
    }

    public void test(View view) {}

    // 点击事件
    public void permissionRequestTest(View view) {
        testRequest();
    }

    // 申请权限  函数名可以随意些
    @Permission(value = Manifest.permission.READ_EXTERNAL_STORAGE, requestCode = 200)
    public void testRequest() {
        Toast.makeText(this, "权限申请成功...", Toast.LENGTH_SHORT).show();
    }

    // 权限被取消  函数名可以随意些
    @PermissionCancel
    public void testCancel() {
        Toast.makeText(this, "权限被拒绝", Toast.LENGTH_SHORT).show();
    }

    // 多次拒绝，还勾选了“不再提示”
    @PermissionDenied
    public void testDenied() {
        Toast.makeText(this, "权限被拒绝(用户勾选了 不再提示)，注意：你必须要去设置中打开此权限，否则功能无法使用", Toast.LENGTH_SHORT).show();
    }
}

File: DzBase/app/src/main/java/com/daipi/dzbase/MainActivity.kt
package com.daipi.dzbase

import android.Manifest
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import com.daipi.base.autoservice.BaseServiceLoader
import com.daipi.base.utils.LogUtil
import com.daipi.common.Cat
import com.daipi.common.autoservice.IPracticeService
import com.daipi.common.autoservice.ITestFmActService
import com.daipi.common.bus.BusKey
import com.daipi.common.bus.TestEvent
import com.daipi.dzbase.databinding.ActivityMainBinding
import com.daipi.dzbase.room.User
import com.daipi.dzbase.viewmodel.MainTestModel
import com.jeremyliao.liveeventbus.LiveEventBus
import dagger.hilt.android.AndroidEntryPoint


import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var cat: Cat

    @Inject
    lateinit var chinaCar: ChinaCar

    @Inject
    lateinit var dog: Dog

    lateinit var bind: ActivityMainBinding

    private val mainModel: MainTestModel by viewModels()

    val MYPERMISSIONS = arrayOf(
        Manifest.permission.READ_PHONE_STATE,
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.INTERNET
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //requestWithPermissions(*MYPERMISSIONS)
        //setContentView(R.layout.activity_main)
        bind = DataBindingUtil.setContentView(this,R.layout.activity_main)

        val userObserver = Observer<User> { user ->
            bind.tvMainApp.text = user.firstName
        }
        mainModel.user.observe(this, userObserver)
/*    override fun getContentViewId(): Int {
        return R.layout.activity_main
    }*/
        registerBus()
    }

    private fun registerBus() {
        LiveEventBus
            .get(TestEvent::class.java)
            .observe(this,
                {
                    LogUtil.d("LiveDataBus ****MainActivity",it)
                    mainModel.user.value = User(null,it.content,it.content) })
    }

    fun toPractice(view: View) {
        //LiveDataBus.get().with("testBus").value = "哈哈哈LiveDataBus"
/*        BaseServiceLoader.load(IPracticeService::class.java)
            ?.toPractice(this, cat.name)*/
        BaseServiceLoader.load(ITestFmActService::class.java)?.toTestFragment(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        LogUtil.d("LiveDataBus ****MainActivity onDestroy()")
    }
}